File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/java/io/github/lishangbu/avalon/dataset/controller/StatControllerTest.java
package io.github.lishangbu.avalon.dataset.controller;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.Stat;
import io.github.lishangbu.avalon.dataset.service.StatService;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class StatControllerTest {
    @Mock private StatService statService;
    @InjectMocks private StatController statController;

    @Test
    void getStatPage_delegatesToService() {
        Stat stat = new Stat();
        Pageable pageable = PageRequest.of(0, 5);
        Page<Stat> page = new PageImpl<>(List.of(stat));
        when(statService.getPageByCondition(any(Stat.class), eq(pageable))).thenReturn(page);

        Page<Stat> result = statController.getStatPage(pageable, stat);

        assertSame(page, result);
        verify(statService).getPageByCondition(stat, pageable);
    }

    @Test
    void listStats_delegatesToService() {
        Stat stat = new Stat();
        List<Stat> list = List.of(stat);
        when(statService.listByCondition(stat)).thenReturn(list);

        List<Stat> result = statController.listStats(stat);

        assertSame(list, result);
        verify(statService).listByCondition(stat);
    }

    @Test
    void save_delegatesToService() {
        Stat stat = new Stat();
        when(statService.save(stat)).thenReturn(stat);

        Stat result = statController.save(stat);

        assertSame(stat, result);
        verify(statService).save(stat);
    }

    @Test
    void update_delegatesToService() {
        Stat stat = new Stat();
        when(statService.update(stat)).thenReturn(stat);

        Stat result = statController.update(stat);

        assertSame(stat, result);
        verify(statService).update(stat);
    }

    @Test
    void deleteById_delegatesToService() {
        statController.deleteById(1L);
        verify(statService).removeById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/kotlin/io/github/lishangbu/avalon/dataset/controller/StatControllerTest.kt
package io.github.lishangbu.avalon.dataset.controller

import io.github.lishangbu.avalon.dataset.entity.dto.SaveStatInput
import io.github.lishangbu.avalon.dataset.entity.dto.StatSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.StatView
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateStatInput
import io.github.lishangbu.avalon.dataset.service.StatService
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

class StatControllerTest {
    @Test
    fun listStats_delegatesToService() {
        val service = FakeStatService()
        val controller = StatController(service)
        val list = listOf(StatView("1", "hp", "生命", 1, false, "2", "physical", "物理"))
        service.listResult = list
        val specification = StatSpecification(id = "1", internalName = "hp", name = "生命", gameIndex = 1, battleOnly = false, moveDamageClassId = "2")

        val result = controller.listStats(specification)

        assertSame(list, result)
        assertEquals("1", service.listCondition!!.id)
    }

    @Test
    fun save_delegatesToService() {
        val service = FakeStatService()
        val controller = StatController(service)
        val command = SaveStatInput("hp", "生命", 1, false, "2")
        service.saveResult = StatView("1", "hp", "生命", 1, false, "2", "physical", "物理")

        val result = controller.save(command)

        assertSame(service.saveResult, result)
        assertSame(command, service.saved)
    }

    @Test
    fun update_delegatesToService() {
        val service = FakeStatService()
        val controller = StatController(service)
        val command = UpdateStatInput("1", "hp", "生命", 1, false, "2")
        service.updateResult = StatView("1", "hp", "生命", 1, false, "2", "physical", "物理")

        val result = controller.update(command)

        assertSame(service.updateResult, result)
        assertSame(command, service.updated)
    }

    @Test
    fun deleteById_delegatesToService() {
        val service = FakeStatService()
        val controller = StatController(service)

        controller.deleteById(1L)

        assertEquals(1L, service.removedId)
    }

    private class FakeStatService : StatService {
        var listCondition: StatSpecification? = null
        var saved: SaveStatInput? = null
        var updated: UpdateStatInput? = null
        var removedId: Long? = null

        var listResult: List<StatView> = emptyList()
        lateinit var saveResult: StatView
        lateinit var updateResult: StatView

        override fun save(command: SaveStatInput): StatView {
            saved = command
            return saveResult
        }

        override fun update(command: UpdateStatInput): StatView {
            updated = command
            return updateResult
        }

        override fun removeById(id: Long) {
            removedId = id
        }

        override fun listByCondition(specification: StatSpecification): List<StatView> {
            listCondition = specification
            return listResult
        }
    }
}
